projects/healthPingValidation.kp/knowledge.md

```diff
@@ -0,0 +1,550 @@
+tldr: Validate incoming data for the new health ping and look at how clients behave.
+thumbnail: images/output_8_0.png
+---
+```python
+import ujson as json
+import matplotlib.pyplot as plt
+import pandas as pd
+import numpy as np
+import plotly.plotly as py
+import pandas as pd
+
+from plotly.graph_objs import *
+from moztelemetry import get_pings_properties, get_one_ping_per_client
+from moztelemetry.dataset import Dataset
+from collections import Counter
+import operator
+
+get_ipython().magic(u'matplotlib inline')
+```
+
+```python
+pings = Dataset.from_source("telemetry") \
+                .where(docType='health', appUpdateChannel="nightly") \
+                .records(sc, sample=1)
+        
+cachedData = get_pings_properties(pings, ["creationDate", "payload/pingDiscardedForSize", "payload/sendFailure", 
+                                             "clientId", "meta/submissionDate", "payload/os", "payload/reason", "application/version"]).cache()
+```
+    fetching 334.53635MB in 88198 files...
+
+
+
+```python
+cachedDataNightly57 = cachedData.filter(lambda ping: ping["meta/submissionDate"] > '20170801').cache()
+```
+# Nightly 56 & 57. Compute failures stats for each failure: sendFailre, discardedForSize. 
+
+* for sendFailure stats include health ping count per failure type
+* for discardedForSize stats include health ping count per ping type
+
+
+```python
+def aggregateFailures(first, second):
+    if first is None:
+        return second
+    if second is None:
+        return first
+    
+    res = first
+    for k, v in second.items():
+        if isinstance(v, int):
+                if k in res:
+                    res[k] += v
+                else: 
+                    res[k] = v;
+    return res
+```
+
+```python
+# return array of pairs [(failureName, {failureStatistic: count, ....}), ...]
+# e.g. [(discardedForSize, {"main": 3, "crash": 5}), (sendFailure, {"timeout" : 34})]
+def getFailuresStatPerFailureName(pings, failureNames):
+    def reduceFailure(failureName):
+        return pings.map(lambda p: p[failureName]).reduce(aggregateFailures)
+    
+    return [(name, reduceFailure(name)) for name in failureNames]
+```
+
+```python
+failuresNames = ["payload/pingDiscardedForSize", "payload/sendFailure"]
+```
+
+```python
+failuresStat = getFailuresStatPerFailureName(cachedData, failuresNames)
+for fs in failuresStat:
+    plt.title(fs[0])
+    plt.bar(range(len(fs[1])), fs[1].values(), align='center')
+    plt.xticks(range(len(fs[1])), fs[1].keys(), rotation=90)
+    plt.show()
+    print fs
+```
+
+
+![png](images/output_8_0.png)
+
+
+    ('payload/pingDiscardedForSize', {u'<unknown>': 4227, u'main': 1563, u'crash': 26, u'bhr': 49})
+
+
+
+
+![png](images/output_8_2.png)
+
+
+    ('payload/sendFailure', {u'eUnreachable': 1736678, u'abort': 6228, u'eChannelOpen': 2962640, u'timeout': 494226})
+
+
+Unknown now represent all oversized pending pings. (https://bugzilla.mozilla.org/show_bug.cgi?id=1384903)
```

Comment:
Instead of "now" use "currently". "Now" makes it sound like it was different before, and _now_ it is the way we want it.